EIPS/eip-9999.md

```diff
@@ -0,0 +1,447 @@
+    # ... existing fields ...
+
+    # Tracks whether CALs were received
+    chunk_access_lists: Dict[Root, Set[ChunkIndex]]
+    # Tracks chunk headers 
+    chunk_headers: Dict[Root, Dict[ChunkIndex, ExecutionChunkHeader]]
+    # Tracks chunk's execution status
+    chunk_execution_statuses: Dict[Root, Dict[ChunkIndex, ExecutionStatus]]
+    # Tracks block validation
+    block_valid: Dict[Root, bool]
+```
+
+#### Three-Phase Validation
+
+##### Phase 1: Forward CALs to EL client
+
+Since CALs are required for validating all subsequent chunks, we forward them to the EL client as soon as they arrive. This avoids sending the same CAL multiple times between CL and EL.
+
+```python
+def on_chunk_access_list_received(store, cal_sidecar):
+    root = cal_sidecar.root
+    chunk_index = cal_sidecar.chunk_index
+
+    # Get or wait for beacon block
+    block = wait_beacon_block(store, root)
+    # Verify sidecar
+    assert verify_chunk_access_list_sidecar(block, cal_sidecar)
+
+    # Send CAL to EL client
+    execution_engine.new_chunk_access_list(block, chunk_index, cal_sidecar.chunk_access_list)
+
+    # Track that CAL was received and unblock any ongoing validation
+    store.chunk_access_lists[root].add(chunk_index)
+    notify_cal_available(root, chunk_index) # unblock wait_cal function
+```
+
+##### Phase 2: Streaming Chunk Validation
+
+Chunks are validated as soon as they and all required CALs are available. 
+
+```python
+def on_chunk_received(store, chunk_sidecar):
+    root = chunk_sidecar.root
+    chunk = chunk_sidecar.chunk
+    chunk_header = chunk.chunk_header
+
+    # Get or wait for beacon block
+    block = wait_beacon_block(store, root)
+    # Verify sidecar
+    assert verify_chunk_sidecar(block, chunk_sidecar)
+
+    # Store chunk header
+    store.chunk_headers[root][chunk_header.index] = chunk_header
+
+    # Wait prior and chunk's CALs
+    for chunk_index in range(chunk_header.index + 1):
+        wait_cal(root, chunk_index)
+
+    # Validate chunk
+    execution_status = execution_engine.validate_chunk(block, chunk) # EL should have all CALs available
+    store.chunk_execution_statuses[root][chunk_header.index] = execution_status
+
+    # Finalize block if all chunks have been validated
+    maybe_finalize_block(store, root)
+```
+
+##### Phase 3: Block finalization
+
+If all chunk and CALs have been received and validated, we can finalize the block.
+
+```python
+def maybe_finalize_block(store, root):
+    if root in store.execution_block_valid:
+        # Block was already finalized
+        return
+    
+    block = get_beacon_block(store, root)
+
+    chunk_headers = store.chunk_headers[root]
+    chunk_statuses = store.chunk_execution_statuses[root]
+
+    chunk_count = len(chunk_statuses)
+    last_chunk_index = chunk_count - 1
+
+    if last_chunk_index not in chunk_headers:
+        # The chunk with last_chunk_index was not validated
+        return
+    if not chunk_headers[last_chunk_index].is_last:
```

Comment:
Do we need `is_last`? Can't we see from gas_used + previous gas used to see this is now the block gas used? So this should be the last chunk?